### PR TITLE
Add option to extract keyframe only during trickplay image generation

### LINKED ISSUE #11336: Use keyframes as scrubbing thumbnails for Trickplay, providing a ~110x speed boost for Trickplay image generation
Currently, trickplay uses the ffmpeg fps filter to extract scrubbing thumbnails.

This is slow as it needs to decode the entire video.
Simply extracting all keyframes instead is roughly 110 times faster (244 vs 2.2 FPS, tested on Win11, Ryzen 7950X3D Eco Mode)

Current ffmpeg command for 10s interval:
```
ffmpeg [...] -i [INPUT] -an -sn -vf "fps=0.10000000149011612,setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709,scale=trunc(min(max(iw\,ih*a)\,240)/2)*2:trunc(ow/a/2)*2,format=yuv420p" -threads 8 -c:v mjpeg -qscale:v 2 -f image2 "[...]/%08d.jpg"
```

My proposal, 110x faster:
```
ffmpeg [...] -skip_frame nokey -i [INPUT] -an -sn -vf "fps=0.10000000149011612,setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709,scale=trunc(min(max(iw\,ih*a)\,240)/2)*2:trunc(ow/a/2)*2,format=yuv420p" -threads 8 -c:v mjpeg -qscale:v 2 -vsync passthrough -f image2 "[...]/%08d.jpg"
```

Note the `-skip_frame nokey` before the input, which tells the decoder to only decode keyframes, and `-vsync passthrough` before the output, which avoids duplicated frames in the output as ffmpeg would otherwise duplicate frames to match the input frame rate.

I'm currently looking at the code and trying to get it to work, though I'm not familiar with the codebase and how the timestamps are stored, if at all.

### PR DESCRIPTION
**Pull request by gnattu**
This would be significantly faster than decoding every frame, but it does have compatibility issues. Not all decoders support this mode, notably the VP9 decoder, CUVID decoders, and QSV decoders.

Some videos with very long key-frame intervals may also perform poorly with this mode, as the image timing could become too inaccurate to reflect the actual frame.

**Changes**

- Add `EnableKeyFrameOnlyExtraction` option to `TrickplayOptions`

**Issues**

Fixes #11336

Web: https://github.com/jellyfin/jellyfin-web/pull/5800

### CODE REVIEW

**Comment by gnattu** [comment]
One more question for the handling: this mode is not compatible with a lot of hardware decoders. Should we add a whitelist to the server and force software decoding when this option is set and current hardware acceleration method is not compatible with this mode?

**Comment by crobibero** [comment]
> Should we add a whitelist to the server and force software decoding when this option is set and current hardware acceleration method is not compatible with this mode?

I think if we know that there is a specific set of configurations that this new setting works with we should add a whitelist.

Also please be sure to PR the docs for next server release

**Comment by nyanmisaka** [comment]
> Not all decoders support this mode, notably the VP9 decoder, CUVID decoders, and QSV decoders.

If FFmpeg's **native** software decoder supports `AVDISCARD_NONKEY`, then its corresponding [hw accelerator](https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/hwaccels.h) should also support it. Hw decoder usually do not, unless they implement it separately.

So the following are exceptions:
- QSV w/ (Prefer OS native DXVA or VA-API hardware decoders)
- NVENC w/ (Enable enhanced NVDEC decoder)
- AMF, uses d3d11va on Windows.

File: MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs [hunk 1]
```diff
@@ -807,12 +807,29 @@ private async Task<string> ExtractImageInternal(string inputPath, string contain
...
             int? qualityScale,
             ProcessPriorityClass? priority,
+            bool enableKeyFrameOnlyExtraction,
             EncodingHelper encodingHelper,
             CancellationToken cancellationToken)
         {
             var options = allowHwAccel ? _configurationManager.GetEncodingOptions() : new EncodingOptions();
             threads ??= _threads;
 
+            if (enableKeyFrameOnlyExtraction)
```
Comment (crobibero) [inline MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs:817]:
```suggestion
            if (allowHwAccel && enableKeyFrameOnlyExtraction)
```

Then the conditional can be simplified a little

**Review by crobibero — APPROVED** [review]

**Review by Bond-009 — APPROVED** [review]

File: MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs [hunk 2]
```diff
@@ -807,12 +807,28 @@ private async Task<string> ExtractImageInternal(string inputPath, string contain
...
             EncodingHelper encodingHelper,
             CancellationToken cancellationToken)
         {
             var options = allowHwAccel ? _configurationManager.GetEncodingOptions() : new EncodingOptions();
             threads ??= _threads;
 
+            if (allowHwAccel && enableKeyFrameOnlyExtraction)
+            {
+                var supportsKeyFrameOnly = (string.Equals(options.HardwareAccelerationType, "nvenc", StringComparison.OrdinalIgnoreCase) && options.EnableEnhancedNvdecDecoder)
+                                           || (string.Equals(options.HardwareAccelerationType, "amf", StringComparison.OrdinalIgnoreCase) && OperatingSystem.IsWindows() && options.PreferSystemNativeHwDecoder)
```
[the remark this answers is not in the document]
Comment (gnattu) [reply MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs:820]: > BTW, how do we handle the VP9 decoder not being able to support the AVDISCARD_NONKEY?

It still works, just as slow as it is not used.

[the remark this answers is not in the document]
Comment (gnattu) [reply MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs]: > Or do we add a whitelist to only allow in video codecs that are known to support this feature? There are some legacy codecs that I haven't checked one by one.

I think just always fallback will be good enough. The legacy codecs are unlikely to be supported by modern hardware anyway.

[the remark this answers is not in the document]
Comment (nyanmisaka) [reply MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs]: OK that makes sense.

**Comment by crobibero** [comment]
Don't forget to make the web change also (or at least create an issue to track it)

**Review by nyanmisaka — APPROVED** [review]

**Pull request merged**